examples/real-world/syncHistoryWithStore.js

```diff
@@ -0,0 +1,103 @@
+export const WILL_NAVIGATE = '@@react-router/WILL_NAVIGATE'
+
+const initialState = {
+  locationBeforeTransitions: null
+}
+
+// Mount this reducer to handle location changes
+export const reducer = (state = initialState, action) => {
+  switch (action.type) {
+    case WILL_NAVIGATE:
+      // Use a descriptive name to make it less tempting to reach into it
+      return {
+        locationBeforeTransitions: action.locationBeforeTransitions
+      }
+    default:
+      return state
+  }
+}
+
+export function syncHistoryWithStore(history, store, {
+  // Specify where you mounted the reducer
+  selectLocationState = state => state.routing,
+  adjustUrlOnReplay = false
+} = {}) {
+  let initialLocation
+  let currentLocation
+  let isTimeTraveling
+  let unsubscribeFromStore
+  let unsubscribeFromHistory
+
+  // What does the store say about current location?
+  const getLocationInStore = (useInitialIfEmpty) => {
+    const locationState = selectLocationState(store.getState())
+    return locationState.locationBeforeTransitions ||
+      (useInitialIfEmpty ? initialLocation : undefined)
+  }
+
+  // Whenever store changes due to time travel, keep address bar in sync
+  const handleStoreChange = () => {
+    const locationInStore = getLocationInStore(true)
+    if (currentLocation === locationInStore) {
+      return
+    }
+
+    // Update address bar to reflect store state
+    isTimeTraveling = true
+    currentLocation = locationInStore
+    history.replace(locationInStore)
```

Comment:
@gaearon Just to finish up this discussion, can you try doing

```
history.transitionTo({ ...locationInStore, action: 'PUSH' })
```

for this line? I believe it will be more correct and will avoid the odd edge cases I mentioned above.

(alternatively, `action: 'REPLACE'` to replace the current history stack frame instead of pushing a new one)